### ISSUE TITLE
[Bug]: aws_elasticache_serverless_cache has too many endpoints

### ISSUE THREAD

**Issue by OJFord** [hashicorp/terraform-provider-aws#39194]
### Terraform Core Version

1.9.5

### AWS Provider Version

5.65.0

### Affected Resource(s)

* aws_elasticache_serverless_cache

### Expected Behavior

`endpoint` should be a single object
`reader_endpoint` should not exist

### Actual Behavior

`endpoint` is a list of nested objects
`reader_endpoint` is a list of nested objects

### Terraform Configuration Files

```tf
resource "aws_elasticache_serverless_cache" "test" {
  engine = "redis"
  name   = "test"
}
```

### Steps to Reproduce

1. create serverless redis cache
2. `terraform console`
3. `>aws_elasticache_serverless_cache.test.endpoint`

### Important Factoids

![image](https://github.com/user-attachments/assets/c52910c5-cb24-4a1d-a38a-923607c324e6)

### References

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/elasticache_serverless_cache#endpoint

https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/GettingStarted.serverless.step2.html ('_the_ Endpoint field')

### Would you like to implement a fix?

None

**Comment by justinretzolk** [comment]
Hey @OJFord 👋 Thank you for taking the time to raise this! In this case, the provider is operating as expected. The schema for the resource matches the what is returned by the [underlying SDK](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/elasticache@v1.40.9/types#ServerlessCache) and [API](https://docs.aws.amazon.com/AmazonElastiCache/latest/APIReference/API_ServerlessCache.html). For what it's worth, it also matches [what's returned by the CLI command](https://docs.aws.amazon.com/cli/latest/reference/elasticache/describe-serverless-caches.html) that's mentioned in the article that you linked.

**Comment by justinretzolk** [comment]
On a second read, I realize I've misread, and you already knew it was an object. Sorry about that! Reopening for further review with that in mind.

**Comment by OJFord** [comment]
@justinretzolk it says they are 'structure' not list of structure?

(It does include reader_endpoint though you're right, is that only for when engine is memcached perhaps?)

**Comment by justinretzolk** [comment]
Looks like we commented at the same time @OJFord 🙂. I'm not entirely certain I know the answers on a quick look, so I'll leave that for someone from the team or community with a bit more context.

**Comment by vtstanescu** [comment]
Found this while looking if there is an open issue for the missing data source for `aws_elasticache_serverless_cache`, and it picked my interest.
Indeed the `endpoint`, and `reader_endpoint`* should be objects, and not a list of objects.

I dug through the commit history and found they are a list of objects since their initial implementation, which, according to the PR (https://github.com/hashicorp/terraform-provider-aws/pull/34951), was implemented by @meetreks and @johnsonaj, so I think they are best to chime in why that's a list and not an object.

\* - `reader_endpoint` is rightly part of the resource as it is part of the ServerlessCache object from AWS API, and AWS SDK for Go v2, see
* https://docs.aws.amazon.com/AmazonElastiCache/latest/APIReference/API_ServerlessCache.html
* https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/elasticache@v1.41.3/types#ServerlessCache

_There is an open issue for what I was actually looking for, sadly ignored for quite some time, https://github.com/hashicorp/terraform-provider-aws/issues/35469_

**Comment by OJFord** [comment]
>  `reader_endpoint` is rightly part of the resource as it is part of the ServerlessCache object from AWS API, and AWS SDK for Go v2

Yeah, I'm not sure if it will have a value for redis (vs. memcached) caches though (if it does it would be the same as the `endpoint` and not read-only) so probably worth documenting as for when `engine = "memcached"`.

https://docs.aws.amazon.com/search/doc-search.html?searchPath=documentation-guide&searchQuery=readerendpoint&this_doc_product=Amazon%20ElastiCache

**Comment by vtstanescu** [comment]
> Yeah, I'm not sure if it will have a value for redis (vs. memcached) caches though (if it does it would be the same as the endpoint and not read-only) so probably worth documenting as for when engine = "memcached".

It does have a value for Redis caches, and it's not the same as the `endpoint` one, the port is different.
Here's a snippet of the AWS CLI response of `aws elasticache describe-serverless-caches --serverless-cache-name $cacheName`

```json
"Engine": "redis",
"Endpoint": {
  "Address": "<redacted>.serverless.euw1.cache.amazonaws.com",
  "Port": 6379
},
"ReaderEndpoint": {
  "Address": "<redacted>.serverless.euw1.cache.amazonaws.com",
  "Port": 6380
},
```

**Comment by ajoga** [comment]
I came here as by reading the documentation I understand that both `endpoint` and `reader_endpoint` are **blocks** and not **lists** of something.

https://github.com/hashicorp/terraform-provider-aws/blob/09dc57c23c298a2236ae3d1128205c7ebb1a870e/website/docs/r/elasticache_serverless_cache.html.markdown#L145-L157

<img width="681" height="982" alt="Image" src="https://github.com/user-attachments/assets/c520fa86-08c7-4a4e-81f5-a1556bf91708" />

**Comment by OJFord** [comment]
@ajoga 'block' in tf docs maps to list/slice of objects in the SDK/Go - e.g. if you configure 2 instances of a block named `foobar` then the provider sees an attr `foobar` with length 2.
